### PR TITLE
fix(web_protocol): replay pipelined requests after failed websocket upgrade

### LINKED ISSUE #12734: Hung connection in failed upgrade edge case
### Describe the bug

This was submitted as a security issue, but after investigation is not a security issue.

The result here is that if a client pipelines another request after a websocket upgrade request, and the upgrade fails in the handler, then the second request gets silently discarded and the client connection will end up hanging.

In reality, a client should never pipeline a second request as they need to wait for the response to know if the second request should be HTTP or WebSocket. So, this is a weird edge case that probably never actually happens.

Will need to figure out a test that reproduces the issue first.

### To Reproduce

Step 1: Initial parse stashes pipelined bytes (web_protocol.py:439-470)

# aiohttp/web_protocol.py:439-466
def data_received(self, data: bytes) -> None:
    if self._force_close or self._close:
        return
    messages: Sequence[_MsgType]
    if self._payload_parser is None and not self._upgraded:
        assert self._parser is not None
        try:
            messages, upgraded, tail = self._parser.feed_data(data)
        except HttpProcessingError as exc:
            messages = [(_ErrInfo(status=400, exc=exc, message=exc.message),
                         EMPTY_PAYLOAD)]
            upgraded = False
            tail = b""

        for msg, payload in messages or ():
            self._request_count += 1
            self._messages.append((msg, payload))

        waiter = self._waiter
        if messages and waiter is not None and not waiter.done():
            waiter.set_result(None)

        self._upgraded = upgraded
        if upgraded and tail:
            self._message_tail = tail        # <-- pipelined bytes parked

On a request that the parser detects as an Upgrade (e.g. Connection: Upgrade + Upgrade: websocket), feed_data returns upgraded=True. Any bytes after the upgrade request's \r\n\r\n come back in tail, which is stashed in self._message_tail for replay after the handler runs.

Step 2: Handler fails the upgrade (web_protocol.py:550-572, web_ws.py:275-319)

The user handler — typically WebSocketResponse.prepare() — raises HTTPBadRequest when handshake validation fails: bad Sec-WebSocket-Version, missing/invalid Sec-WebSocket-Key, bad subprotocol, or any custom validation. The exception is caught in _handle_request:

# aiohttp/web_protocol.py:557-562 (abridged)
except HTTPException as exc:
    resp = exc
    resp, reset = await self.finish_response(request, resp, start_time)

The response is built with default keep_alive=True (only the generic 500/504 path in handle_error sets force_close()), so the connection loop continues.

Step 3: finish_response replays _message_tail and discards the result — THE BUG (web_protocol.py:730-735)

# aiohttp/web_protocol.py:730-735
if self._parser is not None:
    self._parser.set_upgraded(False)
    self._upgraded = False
    if self._message_tail:
        self._parser.feed_data(self._message_tail)      # <-- return value DISCARDED
        self._message_tail = b""

Contrast with data_received at line 447, which captures the same three-tuple messages, upgraded, tail and appends messages to self._messages. Here, the pipelined request inside _message_tail is parsed into a RawRequestMessage + StreamReader — but the tuple is dropped on the floor. The StreamReader is no longer reachable; the message never reaches self._messages; the parser's internal payload state may still be set (if the pipelined request has a body), but no consumer exists to drain it.

Step 4: start() blocks on _waiter forever (web_protocol.py:638-712 — main loop)

# aiohttp/web_protocol.py (abridged main loop)
while not self._force_close:
    if not self._messages:
        try:
            await self._waiter             # <-- blocked
        except asyncio.CancelledError:
            break
    msg, payload = self._messages.popleft()
    ...

_waiter is resolved only when data_received queues a new message. The pipelined message was already consumed (Step 3 dropped it), and data_received will only fire again on new bytes from the peer. A peer that simply stops writing keeps the loop blocked until _keepalive_timeout (default 3630 seconds ≈ 1 hour) closes the connection via _process_keepalive.

### PR DESCRIPTION
**Pull request by goingforstudying-ctrl**
Fixes #12734

When a websocket upgrade fails in the handler and the client has pipelined a second HTTP request, finish_response() previously called self._parser.feed_data(self._message_tail) and discarded the return value. The pipelined request was therefore lost and the connection hung until the keep-alive timeout expired.

Capture the (messages, upgraded, tail) tuple from feed_data(), queue the parsed messages into self._messages, and wake up self._waiter so the main request loop can process the next request immediately.

### CODE REVIEW

File: aiohttp/web_protocol.py [hunk 1]
```diff
@@ -731,8 +731,15 @@ async def finish_response(
             self._parser.set_upgraded(False)
             self._upgraded = False
             if self._message_tail:
-                self._parser.feed_data(self._message_tail)
+                messages, upgraded, tail = self._parser.feed_data(self._message_tail)
                 self._message_tail = b""
+                for msg, payload in messages or ():
```
Comment (Dreamsorcerer) [inline aiohttp/web_protocol.py]: This shouldn't be needed?
```suggestion
                for msg, payload in messages:
```

File: aiohttp/web_protocol.py
Comment (Dreamsorcerer) [inline aiohttp/web_protocol.py]: OK, I think this was needed in an earlier version of the parser, we can just clean it up now.
```suggestion
            for msg, payload in messages:
```

File: tests/test_web_protocol.py [hunk 2]
```diff
@@ -49,3 +51,77 @@ def test_data_received_calls_data_received_cb(
...
+
+async def test_finish_response_replays_empty_message_tail(
+    dummy_manager: Server[BaseRequest],
+) -> None:
+    """No messages queued when parser returns empty list from tail."""
+    event_loop = asyncio.get_running_loop()
+    handler = RequestHandler(dummy_manager, loop=event_loop)
+
+    mock_parser = mock.create_autospec(HttpRequestParser, spec_set=True, instance=True)
+    mock_parser.feed_data.return_value = [], False, b""
```
Comment (Dreamsorcerer) [inline tests/test_web_protocol.py]: I don't think we should be mocking this. At this point I don't think we're really testing anything.

What if we make a variation of the previous test that doesn't have a complete message? I'd assume that would trigger the same result.
`handler._message_tail = b"GET /second HTT"` or something.

**Comment by goingforstudying-ctrl** [comment]
@Dreamsorcerer Thanks for the review! All three suggestions have been applied:

1. ✅ Removed `or ()` fallback in both message loops
2. ✅ Cleaned up the `messages or ()` pattern
3. ✅ Replaced the mock parser test with a real partial message tail test using `b"GET /second HTT"`

Please take another look when you have time.

**Comment by goingforstudying-ctrl** [comment]
Fixed the failing test. The issue was  unconditionally cleared  after feeding it to the parser. When the parser returns empty messages but a non-empty tail (partial data), the tail was lost.

Changed  to  so incomplete messages are preserved for the next iteration.

CI should pass now.

**Comment by goingforstudying-ctrl** [comment]
Fixed the failing test. The issue was finish_response unconditionally cleared _message_tail after feeding it to the parser. When the parser returns empty messages but a non-empty tail (partial data), the tail was lost.

Changed self._message_tail = b"" to self._message_tail = tail so incomplete messages are preserved for the next iteration.

CI should pass now.

**Comment by goingforstudying-ctrl** [comment]
@Dreamsorcerer Fixed. The issue was  clearing the tail unconditionally. Changed to  so partial messages are preserved. CI is green now (except codecov which is expected for a one-line change).

**Comment by Dreamsorcerer** [comment]
Sorry, those tests still aren't testing much. I've redone it myself and found that the uncovered code was unreachable.
#12796
